torch/distributed/_tensor/_utils.py

```diff
@@ -224,3 +224,68 @@ def compute_local_stride(
     return tuple(
         global_stride[i] // stride_divisors[i] for i in range(len(global_stride))
     )
+
+
+def compute_padded_and_unpadded_local_shape(
+    global_shape: ShapeType, mesh: DeviceMesh, placements: Sequence[Placement]
+) -> Tuple[Tuple[int, ...], Tuple[int, ...]]:
+    """
+    This util computes the padded and unpadded local shape of a DTensor. The padded shape is computed by considering
+    applying padding to the global tensor such that each padded shard would have the exact same shape across all ranks.
+    This means sharding happens after padding.
+
+    This differs from `compute_local_shape`. The local shape from `compute_local_shape` considers padding after sharding,
+    meaning padding is applied on each placements instead of globally. Therefore, the local shape on each shard
+    after padding could be different. The local shape returned from `compute_local_shape` is different from the
+    unpadded local shape.
+
+    Padded and unpadded local shape could be the same depending on whether padding is needed on the current shard.
+    """
+
+    # Calculate globally how many chunks a given tensor dim will have globally.
+    num_chunks_by_dim = [1 for _ in enumerate(global_shape)]
+    for mesh_idx, placement in enumerate(placements):
+        if placement.is_shard():
+            tensor_dim = placement.dim  # type: ignore[attr-defined]
+            mesh_dim_size = mesh.size(mesh_idx)
+            num_chunks_by_dim[tensor_dim] *= mesh_dim_size
+
+    full_shard_size, cur_unpadded_shard_size = [], []
+    for size_on_dim, num_chunks in zip(global_shape, num_chunks_by_dim):
+        if num_chunks == 1:
+            # This means no sharding is happening on the ith dimension of the global tensor.
+            # Therefore, the padded and unpadded size of the ith dimension is the same as global_shape[i].
+            full_shard_size.append(size_on_dim)
+            cur_unpadded_shard_size.append(size_on_dim)
```

Comment:
Why we need to have this special casing? I think it could be merged to the else condition

Reply:
The special casing is for if a tensor dim is replicated, then we don't need to go through all the calculation for the full_local_shard_size, current_shard_idx, and current_local_shard_size.